File: HouseRobber.cs
// time complexity = O(n) where n = number of houses
// space complexity = O(1) as we are just using 2 variables
public class Solution
{
    public int Rob(int[] nums)
    {
        int n = nums.Length;
        if (n == 1)
        {
            return nums[0];
        }
        // int[] dp = new int[n];
        // dp = nums;
        // dp[1]= Math.Max(dp[0],dp[1]);
        int prev = nums[0];
        int curr = Math.Max(prev, nums[1]);
        for (int i = 2; i < n; i++)
        {
            //dp[i]=Math.Max(dp[i-1],dp[i]+dp[i-2]);
            int tmp = curr;
            curr = Math.Max(curr, prev + nums[i]);
            prev = tmp;
        }
        return curr;
    }
}

File: House Robber.py
# Time complexity =o(n) and Space Complexity = o(n)
class Solution:
    def rob(self, nums: List[int]) -> int:
        if nums ==0 or len(nums)==0:
            return 
        n= len(nums)
        dp =[[0] * (2 )for j in range (n)]
        dp[0][0]=0
        dp[0][1]=nums[0]

        for i in range(1, n):
            dp[i][0]=max(dp[i-1][0], dp[i-1][1])
            dp[i][1]=nums[i]+dp[i-1][0]

        return max(dp[n-1][0], dp[n-1][1])
        